### PR TITLE
@ #15 | add vmware_provider

### PR DESCRIPTION
**Pull request by hieptranquoc**
connect #15 
pls check this PR!

### CODE REVIEW

File: lib/teracy-dev-core/config/vmware_provider.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    # see: https://www.vagrantup.com/docs/providers/
+    class VmWareProvider < TeracyDev::Config::Configurator
+
+      def configure_node(settings, config)
+        providers_settings = settings['providers'] ||= []
+
+        providers_settings.each do |provider_settings|
+          if provider_settings['type'].include?('vmware')
+            @logger.debug("provider_settings: #{provider_settings}")
+            provider_name = provider_settings['type']
```
Comment (datphan) [inline lib/teracy-dev-core/config/vmware_provider.rb]: `provider_name` is not necessary

File: lib/teracy-dev-core/config/vmware_provider.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+            configure_vmware(provider_settings, config, provider_name) if provider_settings['enabled'] == true
+          end
+        end
+      end
+
+      private
+
+      def configure_vmware(provider_settings, config, provider_name)
```
Comment (datphan) [inline lib/teracy-dev-core/config/vmware_provider.rb]: reductant `provider_name` and `provider_settings` here

File: lib/teracy-dev-core/config/vmware_provider.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        config.vm.provider provider_name do |vmware|
```
Comment (datphan) [inline lib/teracy-dev-core/config/vmware_provider.rb]: should be `config.vm.provider "vmware"`

**Comment by hieptranquoc** [comment]
@datphan I've updated, please check again!

File: lib/teracy-dev-core/config/vmware_provider.rb [hunk 4]
```diff
@@ -0,0 +1,45 @@
+require 'teracy-dev/config/configurator'
+
+module TeracyDevCore
+  module Config
+    # see: https://www.vagrantup.com/docs/providers/
+    class VmWareProvider < TeracyDev::Config::Configurator
```
Comment (hoatle) [inline lib/teracy-dev-core/config/vmware_provider.rb]: VMwareProvider

Comment (hieptranquoc) [reply lib/teracy-dev-core/config/vmware_provider.rb]: @hoatle I fixed!

**Comment by hoatle** [comment]
@datphan please review this asap

**Comment by datphan** [comment]
this doesn't seem to works with this config:
```
default:
  vm:
    networks:
      - _id: "0"
        type: public_network
      - _id: "1"
        type: private_network

  providers:
    - _id: "0"
      enabled: false
      type: virtualbox
    - _id: "1"
      enabled: true
      type: vmware_destop # Another vmware_fusion, vmware_workstation
      gui: false
      clone_directory: "./.vagrant" # Path for storing VMware clones. This value can also be set using the VAGRANT_VMWARE_CLONE_DIRECTORY environment variable. This defaults to ./.vagrant
      enable_vmrun_ip_lookup: true # This defaults to true
      vmx:
        memsize: 1024
        numvcpus: 2

```

Output:
```
k8s-dev datphan$ vagrant up
Bringing machine 'node-01' up with 'virtualbox' provider...
==> node-01: Checking if box 'bento/ubuntu-16.04' is up to date...
==> node-01: Clearing any previously set forwarded ports...
==> node-01: Clearing any previously set network interfaces...
==> node-01: Preparing network interfaces based on configuration...
    node-01: Adapter 1: nat
    node-01: Adapter 2: bridged
    node-01: Adapter 3: hostonly
==> node-01: Forwarding ports...
    node-01: 22 (guest) => 2222 (host) (adapter 1)
==> node-01: Booting VM...
==> node-01: Waiting for machine to boot. This may take a few minutes...
    node-01: SSH address: 127.0.0.1:2222
    node-01: SSH username: vagrant
    node-01: SSH auth method: private key
    node-01: Warning: Connection reset. Retrying...
==> node-01: Machine booted and ready!
==> node-01: Checking for guest additions in VM...
    node-01: The guest additions on this VM do not match the installed version of
    node-01: VirtualBox! In most cases this is fine, but in rare cases it can
    node-01: prevent things such as shared folders from working properly. If you see
    node-01: shared folder errors, please make sure the guest additions within the
    node-01: virtual machine match the version of VirtualBox you have installed on
    node-01: your host and reload your VM.
    node-01: 
    node-01: Guest Additions Version: 5.1.24
    node-01: VirtualBox Version: 5.2
==> node-01: Setting hostname...
==> node-01: Configuring and enabling network interfaces...
==> node-01: Mounting shared folders...
    node-01: /vagrant => /Users/datphan/teracy/k8s-dev
==> node-01: [vagrant-hostmanager:guests] Updating hosts file on active guest virtual machines...
==> node-01: [vagrant-hostmanager:host] Updating hosts file on your workstation (password may be required)...
==> node-01: Machine already provisioned. Run `vagrant provision` or use the `--provision`
==> node-01: flag to force provisioning. Provisioners marked to run always will still run.
==> node-01: Running provisioner: shell...
    node-01: Running: script: Display IP
    node-01: IP Address: 172.28.128.5
```

**Comment by datphan** [comment]
should have docs about this:
```
vagrant plugin install vmware-destop
```

File: config.yaml [hunk 5]
```diff
@@ -45,6 +45,17 @@ default:
       gui: false
       memory: 1024
       description: "%{node_name_prefix} #{Time.now.getutc.to_i}"
+    # - _id: "1"
```
Comment (datphan) [inline config.yaml:59]: lets put some comment about how to get this working:
```
# you need to install vagrant-vmware-desktop first before use this provider
# see the installation steps at https://www.vagrantup.com/docs/vmware/installation.html
```

**Review by datphan — CHANGES_REQUESTED** [review]

File: config.yaml (hunk 5 above)
Comment (hieptranquoc) [reply config.yaml:59]: updated!

**Review by datphan — APPROVED** [review]

File: config.yaml [hunk 6]
```diff
@@ -46,6 +46,20 @@ default:
       memory: 1024
       description: "%{node_name_prefix} #{Time.now.getutc.to_i}"
 
+    # you need to install vagrant-vmware-desktop first before use this provider
+    # see the installation steps at https://www.vagrantup.com/docs/vmware/installation.html
```
Comment (hoatle) [inline config.yaml:57]: or better when you configure the plugins for automation instead of manual

File: config.yaml [hunk 7]
```diff
@@ -46,6 +46,20 @@ default:
...
+    # - _id: "1"
+    #   enabled: true
+    #   type: vmware_destop # Another vmware_fusion, vmware_workstation
```
Comment (hoatle) [inline config.yaml]: type: vmware_desktop

Comment (hoatle) [reply config.yaml]: should support vmware_desktop type only, the 2 others are deprecated so let's not support them.

File: lib/teracy-dev-core/config/vmware_provider.rb [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+  module Config
+    # see: https://www.vagrantup.com/docs/providers/
+    class VMwareProvider < TeracyDev::Config::Configurator
+
+      def configure_node(settings, config)
+        providers_settings = settings['providers'] ||= []
+
+        providers_settings.each do |provider_settings|
+          next if !provider_settings['enabled']
+          if provider_settings['type'].include?('vmware')
```
Comment (hoatle) [inline lib/teracy-dev-core/config/vmware_provider.rb]: be more specific, 'vmware_desktop'

File: lib/teracy-dev-core/config/vmware_provider.rb [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+              provider_settings['vmx'].each do |vmx_key, vmx_val|
+                vmx_val = vmx_val.to_s.strip()
+                if !vmx_val.empty?
+                  vmware.vmx["#{vmx_key}"] = vmx_val
+                end
+              end
+            else
+              val = val.to_s.strip()
+              if !val.empty?
+                vmware.provider_settings[key] = val
```
Comment (hoatle) [inline lib/teracy-dev-core/config/vmware_provider.rb]: the setting like (dynamic key names) this never works, try with other public plugin configs to see:
- https://github.com/teracyhq-incubator/teracy-dev-core/search?q=set_options&unscoped_q=set_options
- https://github.com/hashicorp/vagrant/blob/master/lib/vagrant/plugin/v2/config.rb

**Review by hoatle — CHANGES_REQUESTED** [review]

**Comment by hoatle** [comment]
@hieptranquoc please check my comments

**Comment by hieptranquoc** [comment]
@hoatle I've updated! please check again.

**Review by hoatle — APPROVED** [review]
LGTM, it should work. Bugs will be fixed later if any (as we haven't actually tested with the plugin yet)

**Pull request merged**